src/main/java/com/squareup/kotlinpoet/CodeWriter.kt

```diff
@@ -158,22 +158,44 @@ internal class CodeWriter @JvmOverloads constructor(
    * everywhere else bounds are omitted.
    */
   @Throws(IOException::class)
-  fun emitTypeVariables(typeVariables: List<TypeVariableName>) {
+  fun emitTypeVariables(
+      typeVariables: List<TypeVariableName>,
+      emitSignature: (() -> Unit)? = null) {
     if (typeVariables.isEmpty()) return
 
     emit("<")
+    val withMultipleBounds = mutableListOf<TypeVariableName>()
     var firstTypeVariable = true
     for (typeVariable in typeVariables) {
       if (!firstTypeVariable) emit(", ")
       emitCode("%L", typeVariable.name)
-      var firstBound = true
-      for (bound in typeVariable.bounds) {
-        emitCode(if (firstBound) " : %T" else " & %T", bound)
-        firstBound = false
+      if (typeVariable.bounds.size > 1) {
+        withMultipleBounds += typeVariable
+      } else if (typeVariable.bounds.size == 1) {
+        emitCode(" : %T", typeVariable.bounds[0])
       }
       firstTypeVariable = false
     }
     emit(">")
+    emitSignature?.invoke()
+    if (withMultipleBounds.isNotEmpty()) {
```

Comment:
would this be simpler as two calls? One call `emitTypeVariables()` and another, `emitWhereBlock()` ? Custom control flow is a very powerful construct, and I prefer to use simpler constructs wherever we can